src/main/java/com/example/animalshelter/service/AnimalServiceImpl.java

```diff
@@ -17,10 +19,18 @@ public AnimalServiceImpl(AnimalRepository animalRepository) {
     }
 
     @Override
-    public boolean add(Animal animal) {
+    public AnimalDto add(Animal animal) {
         if (!animalRepository.add(animal)) {
             throw new AddAnimalException(REPOSITORY_EXCEPTION_MSG);
         }
-        return true;
+        return new AnimalDto(animal.getId());
+    }
+
+    @Override
+    public Animal delete(Integer id) throws AnimalNotFoundException {
+        if (id == null || id <= 0) {
+            throw new AnimalNotFoundException(INVALID_ID_MSG);
```

Comment:
Okey I see we are throwing it with different messages, but I am thinking then it this is not different exception causing 400 instead 404 then?

Reply:
What about adding validation on convertions level?
Using `@Valid` annotation and adding `@NotNull` and `@Positive` - which will be your annotation substitute for `@Min(0)`? Generally read about validating `@RequestBody` in spring boot.

Reply:
Finally I have solved this:

 I have add `@NotNull` and `@Positive` annotations (maybe one would be enough) to `@PathVariable`. when parameter is invalid we get `javax.validation.ConstraintViolationException` which I handle in a new class annotated with `@ControllerAdvice` (`AnimalErrorHandler`). This handler method return BAD_REQUEST which is more suitable here as you suggested. For now there is no additional description in this response.

In that case I have removed logic checking id from `AnimalServiceImpl` class and some tests for that scenario too.